### PR TITLE
Add shape context to slicing errors

### PR DESCRIPTION
**Pull request by kklocker**
Hi, just wanted this as I got these errors today and could not figure why this was the case.

Seeing also as this is not an error many people seem to encounter (see picture), I found help difficult to find.
image

If any unit tests are necessary, I would need some help

### CODE REVIEW

**Comment by sklam** [comment]
@kc611, since you are working on advanced indexing, is this something handled from your branch?

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:1659]: I don't think this will print `src_shapes` or `index_shape` in the exception message. `context.call_conv.return_user_exc` will add a constant string (see below) to the LLVM module at compile-time, but both variables are only known at runtime. It will actually print the following:

```python
from numba import njit
import numpy as np

@njit
def bar(a):
    a[:] = np.arange(3)
    return a

a = np.arange(5)
print(bar(a))

ValueError: cannot assign slice of shape [<ir.ExtractValue '.374' of type 'i64', opname 'extractvalue', operands [<ir.InsertValue '.373' of type '[1 x i64]', opname 'insertvalue', operands [<ir.Constant type='[1 x i64]' value=<llvmlite.ir.values._Undefined object at 0x12a41b460>>, <ir.LoadInstr '.372' of type 'i64', opname 'load', operands [<ir.GEPInstr '.371' of type 'i64*', opname 'getelementptr', operands [<ir.GEPInstr '.370' of type '[1 x i64]*', opname 'getelementptr', operands [<ir.AllocaInstr '.341' of type '{i8*, i8*, i64, i64, i64*, [1 x i64], [1 x i64]}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=5>]>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>]>]>] from input of shape (<ir.SelectInstr '.323' of type 'i64', opname 'select', operands [<ir.SelectInstr '.322' of type 'i1', opname 'select', operands [<ir.ICMPInstr '.313' of type 'i1', opname 'icmp', operands [<ir.LoadInstr '.312' of type 'i64', opname 'load', operands [<ir.GEPInstr '.311' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=2>]>]>, <ir.Constant type='i64' value=0>]>, <ir.ICMPInstr '.320' of type 'i1', opname 'icmp', operands [<ir.Instruction '.314' of type 'i64', opname 'sub', operands (<ir.LoadInstr '.310' of type 'i64', opname 'load', operands [<ir.GEPInstr '.309' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=1>]>]>, <ir.LoadInstr '.308' of type 'i64', opname 'load', operands [<ir.GEPInstr '.307' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>)>, <ir.Constant type='i64' value=0>]>, <ir.ICMPInstr '.321' of type 'i1', opname 'icmp', operands [<ir.Instruction '.314' of type 'i64', opname 'sub', operands (<ir.LoadInstr '.310' of type 'i64', opname 'load', operands [<ir.GEPInstr '.309' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=1>]>]>, <ir.LoadInstr '.308' of type 'i64', opname 'load', operands [<ir.GEPInstr '.307' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>)>, <ir.Constant type='i64' value=0>]>]>, <ir.Constant type='i64' value=0>, <ir.Instruction '.319' of type 'i64', opname 'add', operands (<ir.Constant type='i64' value=1>, <ir.Instruction '.318' of type 'i64', opname 'sdiv', operands (<ir.SelectInstr '.317' of type 'i64', opname 'select', operands [<ir.ICMPInstr '.313' of type 'i1', opname 'icmp', operands [<ir.LoadInstr '.312' of type 'i64', opname 'load', operands [<ir.GEPInstr '.311' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=2>]>]>, <ir.Constant type='i64' value=0>]>, <ir.Instruction '.316' of type 'i64', opname 'add', operands (<ir.Instruction '.314' of type 'i64', opname 'sub', operands (<ir.LoadInstr '.310' of type 'i64', opname 'load', operands [<ir.GEPInstr '.309' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=1>]>]>, <ir.LoadInstr '.308' of type 'i64', opname 'load', operands [<ir.GEPInstr '.307' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>)>, <ir.Constant type='i64' value=1>)>, <ir.Instruction '.315' of type 'i64', opname 'sub', operands (<ir.Instruction '.314' of type 'i64', opname 'sub', operands (<ir.LoadInstr '.310' of type 'i64', opname 'load', operands [<ir.GEPInstr '.309' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=1>]>]>, <ir.LoadInstr '.308' of type 'i64', opname 'load', operands [<ir.GEPInstr '.307' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=0>]>]>)>, <ir.Constant type='i64' value=1>)>]>, <ir.LoadInstr '.312' of type 'i64', opname 'load', operands [<ir.GEPInstr '.311' of type 'i64*', opname 'getelementptr', operands [<ir.AllocaInstr '.214' of type '{i64, i64, i64}*', opname 'alloca', operands ()>, <ir.Constant type='i32' value=0>, <ir.Constant type='i32' value=2>]>]>)>)>]>,)
```

To fix that you can use `context.call_conv.return_dynamic_user_exc`:
```python
with builder.if_then(shape_error, likely=False):
    # **Wrongly** assuming that len(src_shapes) == len(index_shape) == 1
    msg = ("cannot assign shape of shape ", src_shapes[0],
            "from input of shape ", index_shape[0])
    nb_types = (None, types.int64, None, types.int64)
    context.call_conv.return_dynamic_user_exc(builder, ValueError,
                                                msg, nb_types)
```

Which will give you the following error message:

```python
$ python repro.py
Traceback (most recent call last):
  File "/Users/guilhermeleobas/git/numba/repro.py", line 117, in <module>
    print(bar(a))
ValueError: ('cannot assign shape of shape ', 3, 'from input of shape ', 5)
```

This is not exactly what you would expect (no string interpolation) but circumvent the issue of raising dynamic exceptions. I'll create an issue for tracking string interpolation inside an exception.

Comment (kklocker) [reply numba/np/arrayobj.py:1659]: Ahh, thanks for clarifying :) Ill see if I can get to it soon enough

Comment (guilhermeleobas) [reply numba/np/arrayobj.py:1659]: Let me know if you need any help. I think the interface for `call_conv.return_dynamic_user_exc` could be improved.

Comment (kklocker) [reply numba/np/arrayobj.py:1659]: @guilhermeleobas For the case above here, we don't know that it is the first index that causes the trouble. Is there a type for tubles / lists in the lowering I can use here?

I tried `Tuple`, `List`, and also flattening `*[types.int32 for _ in src_shapes]`
but it does not seem to work
```python
 with builder.if_then(shape_error, likely=False):
            msg = ("cannot assign slice of shape ", src_shapes,
                   "from input of shape ", index_shape)

            nb_types = (None, types.Tuple, None, types.Tuple)
            context.call_conv.return_dynamic_user_exc(builder, ValueError, msg,
                                                       nb_types)
```

Comment (kklocker) [reply numba/np/arrayobj.py]: I think I found an ok compromise:

```python
# repro.py
from numba import njit
import numpy as np

@njit
def bar(a):
    a[:] = np.arange(3)
    return a
a = np.arange(12).reshape((3, 2, 2))

print(bar(a))
# ValueError: ('cannot assign slice of shape', 3, 2, 3, 'from input of shape', 3, 2, 2)
```

File: numba/np/arrayobj.py
Comment (guilhermeleobas) [inline numba/np/arrayobj.py:1700]: There's a typo here:
```suggestion
            context.call_conv.return_dynamic_user_exc(builder, ValueError, msg, nb_types)
```

File: numba/np/arrayobj.py
Comment (kklocker) [reply numba/np/arrayobj.py]: Nice, thank you! 
Implemented -- although I have not been able to test the Sequence - part of this flow, i don't know how to reproduce this error (or even how to construct a numba Sequence ....)

**Comment by kklocker** [comment]
... Getting a wils spread of test failures / passes here now
https://dev.azure.com/numba/numba/_build/results?buildId=15627&view=results
image

**Comment by guilhermeleobas** [comment]
Hi @kklocker, to address the CI failures, you need to update the exception messages in the tests as well.

**Comment by kklocker** [comment]
Fixed the assertion messages (although I don't really understand why one should assert equality on the message itself and not just the error(+type)) 

Fails on `assert_no_memory_leak` now🤷Don't really see how this is related

https://dev.azure.com/numba/numba/_build/results?buildId=15651&view=results
image

**Comment by gmarkall** [comment]
Added "Waiting on reviewer" pending advice on the `assert_no_memory_leak` issue mentioned above.

**Comment by guilhermeleobas** [comment]
https://github.com/numba/numba/pull/9262 shall fix the CI error:

```
  File "/home/vsts/work/1/s/numba/np/arrayobj.py", line 554, in setitem_array
    return fancy_setslice(context, builder, sig, args,
  File "/home/vsts/work/1/s/numba/np/arrayobj.py", line 1672, in fancy_setslice
    context.call_conv.return_dynamic_user_exc(builder, ValueError,
  File "/home/vsts/work/1/s/numba/core/callconv.py", line 722, in return_dynamic_user_exc
    self.set_dynamic_user_exc(builder, exc, exc_args, nb_types,
  File "/home/vsts/work/1/s/numba/core/callconv.py", line 689, in set_dynamic_user_exc
    struct_gv = builder.load(pyapi.serialize_object(exc))
  File "/home/vsts/work/1/s/numba/core/pythonapi.py", line 1412, in serialize_object
    struct = self.serialize_uncached(obj)
  File "/home/vsts/work/1/s/numba/core/pythonapi.py", line 1380, in serialize_uncached
    data = serialize.dumps(obj)
  File "/home/vsts/work/1/s/numba/core/serialize.py", line 58, in dumps
    p.dump(obj)
  File "/home/vsts/work/1/s/numba/cloudpickle/cloudpickle_fast.py", line 644, in dump
    raise pickle.PicklingError(msg) from e
_pickle.PicklingError: Could not pickle object as excessively deep recursion required.
```

**Comment by guilhermeleobas** [comment]
@kklocker, could you copy/paste the changes done on `callconv.py` in https://github.com/numba/numba/pull/9262 to this PR. It should fix the current CI failure.

**Comment by kklocker** [comment]
@guilhermeleobas Cherry-picked your commit :)

**Review by guilhermeleobas — APPROVED** [review]

**Comment by stuartarchibald** [comment]
Thanks for implementing this @kklocker and for reviewing this @guilhermeleobas . Waiting on another review as this PR contains a patch from #9262 which is under review and authored by the reviewer of this PR! Seems like this PR might be a way to trigger the issue fixed by #9692 which will aid in understanding the problem.

**Comment by guilhermeleobas** [comment]
When constructing the exception info structure, to know whether a variable is static or dynamic is determined by its type within `exc_args`. For instance, consider the statement raise `ValueError('static arg', x)`, where `x` is a string. The representation of this in `exc_args` has been simplified to the following:

```python
('static arg', <ir.InsertValue 'inserted.parent.2' ...>)
```

The patch I suggest simplifies the process of represent `ir.Value` nodes, as they are only used to indicate a variable is static or dynamic. The patch simply replaces those notes by a placeholder value.

**Comment by stuartarchibald** [comment]
> @stuartarchibald, when constructing the exception info structure, to know whether a variable is static or dynamic is determined by its type within `exc_args`. For instance, consider the statement raise `ValueError('static arg', x)`, where `x` is a string. The representation of this in `exc_args` has been simplified to the following:
> 
> ```python
> ('static arg', <ir.InsertValue 'inserted.parent.2' ...>)
> ```
> 
> The patch I suggest simplifies the process of represent `ir.Value` nodes, as they are only used to indicate a variable is static or dynamic. The patch simply replaces those notes by a placeholder value.

@guilhermeleobas thanks for this explanation, it provides useful context when reviewing. Do you recall which tests were failing without this patch?

**Comment by guilhermeleobas** [comment]
@stuartarchibald, `test_cross` in `test_np_functions.py`. To reproduce, undo the changes in `callconv.py` and run the command below:

```
./runtests.py -v -f numba.tests.test_np_functions.TestNPFunctions -k test_cross
```

**Comment by guilhermeleobas** [comment]
Hi, I think other core devs are a bit busy with upcoming SciPy conf that is happening this week. Hopefully this can be reviewed again soon.

**Review by sklam — APPROVED** [review]
I verified the pickle error fix

**Comment by sklam** [comment]
Btw, the reproducer drifted. I have to use `numba.tests.test_np_functions.TestNPFunctions.test_unwrap_basic`.

**Pull request merged**
